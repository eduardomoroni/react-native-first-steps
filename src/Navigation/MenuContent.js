// @flow

import React, { Component } from 'react'
import { ScrollView, Image, BackAndroid } from 'react-native'
import { connect } from 'react-redux'
import styles from './Styles/MenuContentStyle'
import { Images } from '../Themes'
import { userLoggout } from '../Redux/Actions'
import DrawerButton from '../Containers/Components/MenuButton'
import { Actions as NavigationActions } from 'react-native-router-flux'
import I18n from 'react-native-i18n'

class DrawerContent extends Component {
  componentDidMount () {
    BackAndroid.addEventListener('hardwareBackPress', () => {
      if (this.context.drawer.props.open) {
        this.toggleDrawer()
        return true
      }
      return false
    })
  }

  toggleDrawer () {
    this.context.drawer.toggle()
  }

  handlePress = (route: string) => {
    this.toggleDrawer()
    switch (route) {
      case 'settings':
        NavigationActions.settings()
        break
      case 'cardSearch':
        NavigationActions.cardSearch()
        break
      case 'loginScreen':
        NavigationActions.loginScreen()
        break
    }
  }

  renderLoginOrLogout = () => {
    if (this.props.user === null) {
      return (
        <DrawerButton text={I18n.t('login')} onPress={() => this.handlePress('loginScreen')} />
      )
    } else {
      return (
        <DrawerButton text={I18n.t('logout')} onPress={() => this.props.onLogout()} />
      )
    }
  }

  render () {
    return (
      <ScrollView style={styles.container}>
        <Image source={Images.logo} style={styles.logo} />
        <DrawerButton text={I18n.t('settings')} onPress={() => this.handlePress('settings')} />
        {this.renderLoginOrLogout()}
      </ScrollView>
    )
  }
}

DrawerContent.contextTypes = {
  drawer: React.PropTypes.object
}

const mapStateToProps = (state) => {
  const { user } = state.user
  return {
    user: user
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onLogout: () => dispatch(userLoggout())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(DrawerContent)
