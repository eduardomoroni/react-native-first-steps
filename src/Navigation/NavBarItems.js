/* @flow */

import React from 'react'
import {
  StyleSheet,
  TouchableOpacity
} from 'react-native'
import { Colors, Metrics } from '../Styles/Themes/'
import Icon from 'react-native-vector-icons/FontAwesome'
import { navButtonColor } from '../Styles/NavBarStyle'

type ButtonPropsType = {
  name: string,
  size: number,
  color: string,
  style: any
}

export default {
  navButton (buttonProps: ButtonPropsType, callback: Function) {
    return (
      <TouchableOpacity onPress={callback}>
        <Icon {...defaultButtonProps} {...buttonProps} />
      </TouchableOpacity>
    )
  }
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: Colors.transparent,
    color: navButtonColor,
    justifyContent: 'center'
  }
})

const defaultButtonProps: ButtonPropsType = {
  name: 'heart',
  size: Metrics.icons.medium,
  color: Colors.snow,
  style: styles.button
}
