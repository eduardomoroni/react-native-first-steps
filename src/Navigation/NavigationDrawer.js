// @flow

import React, { PropTypes, Component } from 'react'
import Drawer from 'react-native-drawer'
import { DefaultRenderer, Actions as NavigationActions } from 'react-native-router-flux'
import Menu from './NavigationMenu'
import { connect } from 'react-redux'
import { drawerStyle } from '../Styles/NavBarStyle'

export const menuDrawerProps = {
  key: 'drawer',
  open: false,
  side: 'left',
  type: 'displace',
  content: <Menu />
}

class NavigationDrawer extends Component {
  render () {
    const state = this.props.navigationState
    const defaultProps = {
      ref: 'navigation',
      open: state.open,
      onOpen: () => NavigationActions.refresh({key: state.key, open: true, side: state.side}),
      onClose: () => NavigationActions.refresh(menuDrawerProps),
      content: state.content,
      styles: drawerStyle,
      tapToClose: true,
      openDrawerOffset: 0.25,
      panCloseMask: 0.25,
      negotiatePan: true,
      type: 'overlay',
      side: state.side, // If we set dynamically to right we're getting error
      tweenHandler: tweenHandler
    } // I need to find a better way to override this props in execution time

    return (
      <Drawer {...defaultProps}>
        <DefaultRenderer navigationState={state.children[0]} onNavigate={this.props.onNavigate} />
      </Drawer>
    )
  }
}

const tweenHandler = (ratio) => {
  return {
    main: {
      opacity: Math.max(0.54, 1 - ratio) }
  }
}

NavigationDrawer.propTypes = {
  navigationState: PropTypes.object,
  onNavigate: PropTypes.func
}

export default connect()(NavigationDrawer)
