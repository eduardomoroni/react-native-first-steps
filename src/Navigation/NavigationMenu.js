// @flow

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { ScrollView, Image, BackAndroid } from 'react-native'
import { connect } from 'react-redux'
import { Images } from '../Styles/Themes'
import { userLoggout } from '../Redux/Actions'
import { DrawerButton } from '../Containers/Components'
import { Actions as NavigationActions } from 'react-native-router-flux'
import I18n from 'react-native-i18n'
import type { Dispatch } from 'redux'

class NavigationMenu extends Component {
  componentDidMount () {
    BackAndroid.addEventListener('hardwareBackPress', () => {
      if (this.context.drawer.props.open) {
        this.toggleDrawer()
        return true
      }
      return false
    })
  }

  toggleDrawer () {
    this.context.drawer.toggle()
  }

  handlePress = (route: string) => {
    this.toggleDrawer()
    NavigationActions[route]()
  }

  renderLoginOrLogout = () => {
    if (this.props.user === null) {
      return (
        <DrawerButton text={I18n.t('login')} onPress={() => this.handlePress('loginScreen')} />
      )
    } else {
      return (
        <DrawerButton text={I18n.t('logout')} onPress={() => this.props.onLogout()} />
      )
    }
  }

  render () {
    return (
      <ScrollView style={styles.container}>
        <Image source={Images.logo} style={styles.logo} />
        <DrawerButton text={I18n.t('cardSearch')} onPress={() => this.handlePress('cardSearchForm')} />
        <DrawerButton text={I18n.t('wishList')} onPress={() => this.handlePress('wishList')} />
        <DrawerButton text={I18n.t('settings')} onPress={() => this.handlePress('settings')} />
        {this.renderLoginOrLogout()}
      </ScrollView>
    )
  }
}

NavigationMenu.contextTypes = {
  drawer: PropTypes.object
}

NavigationMenu.propTypes = {
  user: PropTypes.object,
  onLogout: PropTypes.func
}

const mapStateToProps = (state) => {
  const { token } = state.auth

  return {
    user: token
  }
}

const mapDispatchToProps = (dispatch: Dispatch<*>) => {
  return {
    onLogout: () => dispatch(userLoggout())
  }
}

const styles = {
  container: {
    flex: 1,
    padding: 20
  },
  logo: {
    alignSelf: 'center'
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(NavigationMenu)
