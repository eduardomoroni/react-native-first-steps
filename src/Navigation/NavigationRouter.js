// @flow

import React from 'react'
import { Scene, Router } from 'react-native-router-flux'
import Settings from '../Containers/Settings'
import LoginScreen from '../Containers/LoginScreen'
import MainScreen from '../Containers/MainScreen'
import Styles from '../Styles/NavigationRouterStyle'
import Drawer from './NavigationDrawer'
import I18n from 'react-native-i18n'

const NavigationRouter = () => {
  return (
    <Router>
      <Scene key='drawer' component={Drawer} open={false} >
        <Scene key='drawerChildrenWrapper' navigationBarStyle={Styles.navBar} titleStyle={Styles.title} leftButtonIconStyle={Styles.leftButton} rightButtonTextStyle={Styles.rightButton} >
          <Scene initial key='MainScreen' component={MainScreen} title={I18n.t('welcome')} />
          <Scene key='settings' component={Settings} title={I18n.t('settings')} />
          <Scene key='loginScreen' component={LoginScreen} title={I18n.t('login_for_title')} hideNavBar />
        </Scene>
      </Scene>
    </Router>
  )
}

export default NavigationRouter
