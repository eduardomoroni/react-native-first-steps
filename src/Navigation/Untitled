{
    "name": "two",
    "version": "1.0.0",
    "main": "index.js",
    "license": "MIT",
    "scripts": {
        "echo": "echo \"ONE\" && exit 0"
    }
}
