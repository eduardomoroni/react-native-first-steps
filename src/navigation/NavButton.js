import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { TouchableOpacity } from 'react-native'
import Icon from 'react-native-vector-icons/FontAwesome'
import { Metrics } from '../styles/themes/'
import styles from '../styles/navigation/NavBarItemsStyle'

export class NavButton extends Component {
  render () {
    const {
      iconProps,
      callback
    } = this.props

    return (
      <TouchableOpacity onPress={callback} >
        <Icon {...iconProps} />
      </TouchableOpacity>
    )
  }
}

NavButton.propTypes = {
  iconProps: PropTypes.shape({
    name: PropTypes.string,
    size: PropTypes.number,
    style: PropTypes.any
  }),
  callback: PropTypes.func
}

Icon.defaultProps = {
  name: 'heart',
  size: Metrics.icons.medium,
  style: styles.button
}
