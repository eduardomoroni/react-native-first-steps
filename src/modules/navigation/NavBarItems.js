import React from 'react'
import I18n from 'react-native-i18n'
import { Text, View } from 'react-native'
import { Actions as NavigationActions } from 'react-native-router-flux'
import { builtInBarStyle, customBarStyle } from '../theme/NavBarStyle'
import { NavButton } from './NavButton'

type ButtonPropsType = {
  name: string,
  size: number,
  style: any
}

// TODO: Enable Flow
// TODO: REFACTOR TYPES
export const navButton = (buttonProps: ButtonPropsType, callback: Function, key: any) => {
  return (
    <NavButton iconProps={{...buttonProps}} callback={callback} key={key} />
  )
}

export const openDrawer = (component: any) => {
  NavigationActions.refresh({
    key: 'drawer',
    content: component,
    open: true,
    side: 'left'
  })
}

export const navTitle = (title: string) => {
  return (
    <Text style={builtInBarStyle.titleStyle}>
      {I18n.t(title)}
    </Text>
  )
}

// TODO: This is an array of Object ButtonPropsType
export const navButtonBar = (buttons: any) => {
  return (
    <View style={[customBarStyle.rightButton, customBarStyle.buttons]} >
      { buttons.map(i => navButton(i.buttonProps, i.callback, i.buttonProps.name)) }
    </View>
  )
}

export const navBackButton = () => {
  return navButton({name: 'angle-left'}, NavigationActions.pop)
}
