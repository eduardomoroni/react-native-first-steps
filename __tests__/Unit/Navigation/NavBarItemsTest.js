import ReactNative from 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import * as NavItems from '../../../src/Navigation/NavBarItems'
import { NavButton } from '../../../src/Navigation/NavButton'
import { Actions as NavigationActions } from 'react-native-router-flux'

const props = {
  iconProps: {
    name: 'search-plus',
    size: 30,
    style: {}
  },
  callback: jest.fn()
}

describe('NavBarItems Tests', () => {
  it('should render <NavButton />', () => {
    const defaultButton = NavItems.navButton(props.iconProps, props.callback)
    expect(defaultButton).toEqual(<NavButton {...props} />)
  })

  it('should expose openDrawer callback', () => {
    const expectedPayload = {
      key: 'drawer',
      content: ReactNative.View,
      open: true,
      side: 'left'
    }

    NavItems.openDrawer(ReactNative.View)
    expect(NavigationActions.refresh).toHaveBeenCalledWith(expectedPayload)
  })

// Warning: Unknown props `accessible`, `allowFontScaling`, `ellipsizeMode` on <Text> tag. Remove these props from the element. For details, see https://fb.me/react-unknown-prop
  it('should render navTitle', () => {
    const wrapper = shallow(NavItems.navTitle('TITLE'))
    expect(wrapper).toMatchSnapshot()
  })

  it('should render multiples navButtons on right', () => {
    const buttons = [{
      buttonProps: {name: 'sliders'},
      callback: jest.fn()
    }, {
      buttonProps: {name: 'close'},
      callback: jest.fn()
    }]

    const wrapper = shallow(NavItems.navButtonBar(buttons))
    expect(wrapper).toMatchSnapshot()
  })

  it('should render navBackButton', () => {
    const wrapper = shallow(NavItems.navBackButton())
    expect(wrapper).toMatchSnapshot()
    wrapper.simulate('press')
    expect(NavigationActions.pop).toHaveBeenCalled()
  })
})


