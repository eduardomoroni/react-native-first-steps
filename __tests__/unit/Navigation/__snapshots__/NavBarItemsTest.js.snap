// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBarItems Tests should render multiples navButtons on right 1`] = `
<View
  style={
    Array [
      Object {
        "justifyContent": "flex-end",
      },
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      },
    ]
  }
>
  <NavButton
    callback={[Function]}
    iconProps={
      Object {
        "name": "sliders",
      }
    }
  />
  <NavButton
    callback={[Function]}
    iconProps={
      Object {
        "name": "close",
      }
    }
  />
</View>
`;

exports[`NavBarItems Tests should render navBackButton 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  focusedOpacity={0.7}
  onPress={[Function]}
>
  <Icon
    name="angle-left"
    size={30}
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0)",
        "color": "white",
        "justifyContent": "center",
        "marginLeft": 5,
      }
    }
  />
</TouchableOpacity>
`;

exports[`NavBarItems Tests should render navTitle 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "white",
      "fontFamily": "beleren",
      "fontSize": 18,
    }
  }
>
  I18n.t(TITLE)
</Text>
`;
